DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,127 @@
+[Expression-bodied members](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members).
+This feature has been available for use since either C# version 6 or 7 depending on context - for reference the most recent stable release is version 9.0.
+
+The C# documentation linked above gives an example of idiomatic use of this feature (reproduced below). In considering this piece of code, we 
+should be careful to remember that C# (to some degree of approximation at least) encourages access to state through methods (Getters and Setters, properties, etc.),
+so the use of this feature may be more profitable in C# than in D where the aforementioned idiom is not as popular - that is to say that in the authors experience 
+D programs with many single line functions bodies are significantly less prevalent than in C#.
+```c#
+public class Location
+{
+   private string locationName;
+
+   public Location(string name)
+   {
+      locationName = name;
+   }
+
+   public string Name => locationName;
+}
+```
+## Description
+### Syntax 
+The grammar changes are as follows:
+```diff
+FunctionBody:
+     SpecifiedFunctionBody
+     MissingFunctionBody
++    ShortenedFunctionBody
+...
+
++ShortenedFunctionBody:
++    => AssignExpression ;
+```
+### Semantics 
+This proposed semantics are a simple example of what is often referred to as lowering:
+```
+FunctionDeclarator => AssignExpression;
+```
+shall be rewritten to
+```
+FunctionDeclarator
+{
+	return AssignExpression;
+}
+```
+Although valid syntax, this construct is semantically nonsense in the case of a constructor or destructor (i.e. they cannot have a return value) and 
+use in these contexts should be explicitly rejected early by the implementation to ensure a meaningful error message for the programmer. 
+
+The crux of the [implementation](https://github.com/dlang/dmd/blob/master/src/dmd/parse.d#L5139) is listed with annotations below:
+```D
+const returnloc = token.loc;
+nextToken();//Walk past the =>
+// ↓ the where the lowered result goes     |  ↓ Ingest the expression following the =>
+f.fbody = new AST.ReturnStatement(returnloc, parseExpression());
+f.endloc = token.loc;
+check(TOK.semicolon);
+```
+Although this listing is informative, the *location* of this implementation within the parser (within `parseContracts`) should not go unquestioned (elaborated in the following section). It should 
+also be noted that the above code uses `parseExpression` rather than parsing an assign expression specifically - the author does not know if this matters in practice (or as an exercise).
+
+## Reference
+Recently, [it has been raised](https://github.com/dlang/dlang.org/pull/3059) that the implementation currently allows the use of function contracts with 
+this syntax. This is not possible with function literals, so it has not been included in the body of this DIP however it is mentioned here both for future 
```

Comment:
Function literals do support contracts:

https://github.com/dlang/dmd/blob/master/test/compilable/extra-files/header1.d#L17-L31

It's just the shortened versions which do not.

It might actually make more sense to fix this in the other direction, and allow contracts on either syntax (though currently the syntax does have [some oddities](https://github.com/dlang/dlang.org/pull/3059#issuecomment-877681549).

Reply:
Shortened syntax should have minimum inconsistency with normal function declarations. That's why the implementation is where it is, so it hooks into all that existing machinery and allows the same things. (or most the same things, i never tried it a short body with a long contract...)

Reply:
Yes, long contracts seem to only be allowed if they're not the last contract, which is the oddity mentioned above.

Reply:
> Function literals do support contracts:
> 
> https://github.com/dlang/dmd/blob/master/test/compilable/extra-files/header1.d#L17-L31
> 
> It's just the shortened versions which do not.
> 
> It might actually make more sense to fix this in the other direction, and allow contracts on either syntax (though currently the syntax does have [some oddities](https://github.com/dlang/dlang.org/pull/3059#issuecomment-877681549).

GOOD point (The latter point will have to be for a later review I think)